pkg/ingester/streams_map_test.go

```diff
@@ -0,0 +1,93 @@
+package ingester
+
+import (
+	"testing"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/loki/pkg/validation"
+)
+
+func TestStreamsMap(t *testing.T) {
+	limits, err := validation.NewOverrides(defaultLimitsTestConfig(), nil)
+	require.NoError(t, err)
+	limiter := NewLimiter(limits, NilMetrics, &ringCountMock{count: 1}, 1)
+
+	ss := []*stream{
+		newStream(
+			defaultConfig(),
+			limiter,
+			"fake",
+			model.Fingerprint(1),
+			labels.Labels{
+				{Name: "foo", Value: "bar"},
+			},
+			true,
+			NilMetrics,
+		),
+		newStream(
+			defaultConfig(),
+			limiter,
+			"fake",
+			model.Fingerprint(2),
+			labels.Labels{
+				{Name: "bar", Value: "foo"},
+			},
+			true,
+			NilMetrics,
+		),
+	}
+	var s *stream
+	var loaded bool
+
+	streams := newStreamsMap()
+
+	require.Equal(t, 0, streams.Len())
+	s, loaded = streams.Load(ss[0].labelsString)
+	require.Nil(t, s)
+	require.False(t, loaded)
+	s, loaded = streams.LoadByFP(ss[1].fp)
+	require.Nil(t, s)
+	require.False(t, loaded)
+
+	s, loaded, err = streams.LoadOrStoreNew(ss[0].labelsString, func() (*stream, error) {
+		return ss[0], nil
+	})
+	require.Equal(t, s, ss[0])
+	require.False(t, loaded)
+	require.Nil(t, err)
+
+	s, loaded, err = streams.LoadOrStoreNewByFP(ss[1].fp, func() (*stream, error) {
+		return ss[1], nil
+	})
+	require.Equal(t, s, ss[1])
+	require.False(t, loaded)
+	require.Nil(t, err)
+
+	require.Equal(t, len(ss), streams.Len())
```

Comment:
super tiny nit: since this test is making sure both the `LoadOrStoreNewByFP` calls worked, and thus not directly related to `len(ss)`, I would test against the hard-coded 2 here just so the reader doesn't have to keep that extra context of how big `ss` is.

```suggestion
	require.Equal(t, 2, streams.Len())
```

Reply:
I believe this makes more sense as it is, because we can safely add more streams to the test case and this will reflect the expectation that they're all stored. The prior checks could more idiomatically be run in a loop over each stream, but that's a tiny suggestion.

Reply:
yeah, I guess my thought is that adding more cases to the slice won't actually cause them to be stored, because it's a function of actually calling the `LoadOrStore`. However, if we put that in a loop like you mentioned then I can see this argument. But now we're splitting hairs, doesn't really matter either way.